modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java

```diff
@@ -117,119 +120,103 @@ private CompletableFuture<?> onSchemaChange(ConfigurationNotificationEvent<Named
-
-                createSchemaFut.thenCompose(t -> metastorageMgr.invoke(
-                        Conditions.notExists(key),
-                        Operations.put(key, SchemaSerializerImpl.INSTANCE.serialize(schemaDescFromUpdate)),
-                        Operations.noop()));
-            } catch (Throwable th) {
-                createSchemaFut.completeExceptionally(th);
-            }
-
-            createSchemaFut.whenComplete((ignore, th) -> {
-                if (th == null) {
-                    registriesVv.get(causalityToken).thenRun(() -> inBusyLock(busyLock,
-                            () -> fireEvent(SchemaEvent.CREATE, new SchemaEventParameters(causalityToken, tblId, schemaDescFromUpdate))));
+            // Fire event early, because dependent listeners have to register VersionedValues' update futures
+            var eventParams = new SchemaEventParameters(causalityToken, tblId, newSchema);
+
+            fireEvent(SchemaEvent.CREATE, eventParams)
+                    .whenComplete((v, e) -> {
+                        if (e != null) {
+                            LOGGER.warn("Error when processing CREATE event", e);
+                        }
+                    });
+
+            return registriesVv.update(causalityToken, (registries, e) -> inBusyLock(busyLock, () -> {
+                if (e != null) {
+                    return failedFuture(new IgniteInternalException(IgniteStringFormatter.format(
+                            "Cannot create a schema for the table [tblId={}, ver={}]", tblId, newSchemaVersion), e)
+                    );
                 }
-            });
 
-            return createSchemaFut;
+                return registerSchema(registries, tblId, tblCfg.name(), newSchema);
+            }));
         } finally {
             busyLock.leaveBusy();
         }
     }
 
-    /**
-     * Create new schema locally.
-     *
-     * @param causalityToken Causality token.
-     * @param tableId Table id.
-     * @param tableName Table name.
-     * @param schemaDescriptor Schema descriptor.
-     * @return Create schema future.
-     */
-    private CompletableFuture<?> createSchema(long causalityToken, UUID tableId, String tableName, SchemaDescriptor schemaDescriptor) {
-        if (!busyLock.enterBusy()) {
-            throw new IgniteException(new NodeStoppingException());
+    private void setColumnMapping(SchemaDescriptor schema, UUID tableId) throws ExecutionException, InterruptedException {
+        if (schema.version() == INITIAL_SCHEMA_VERSION) {
+            return;
         }
 
-        try {
-            return createSchemaInternal(causalityToken, tableId, tableName, schemaDescriptor);
-        } finally {
-            busyLock.leaveBusy();
+        int prevVersion = schema.version() - 1;
+
+        SchemaDescriptor prevSchema = searchSchemaByVersion(tableId, prevVersion);
+
+        if (prevSchema == null) {
+            prevSchema = schemaByVersion(tableId, prevVersion).get();
```

Comment:
Yes, I left a comment, why this blocking call is unavoidable, because it is used in the synchronous part of the a configuration listener. Or do you think we can get rid of it?

Reply:
It seems to me that we cant, as we need the schema descriptor to fire event, which should be fired in synchronous part of the listener.